ui/packages/consul-ui/app/components/token-source/index.js

```diff
@@ -1,36 +1,47 @@
-import Component from '@ember/component';
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import { tracked } from '@glimmer/tracking';
 
 import chart from './chart.xstate';
-export default Component.extend({
-  onchange: function() {},
-  init: function() {
-    this._super(...arguments);
+
+export default class TokenSource extends Component {
+  @tracked provider;
+  @tracked jwt;
+
+  constructor() {
+    super(...arguments);
     this.chart = chart;
-  },
-  actions: {
-    isSecret: function() {
-      return this.type === 'secret';
-    },
-    change: function(e) {
-      e.data.toJSON = function() {
-        return {
-          AccessorID: this.AccessorID,
-          // TODO: In the past we've always ignored the SecretID returned
-          // from the server and used what the user typed in instead, now
-          // as we don't know the SecretID when we use SSO we use the SecretID
-          // in the response
-          SecretID: this.SecretID,
-          Namespace: this.Namespace,
-          ...{
-            AuthMethod: typeof this.AuthMethod !== 'undefined' ? this.AuthMethod : undefined,
-            // TODO: We should be able to only set namespaces if they are enabled
-            // but we might be testing for nspaces everywhere
-            // Namespace: typeof this.Namespace !== 'undefined' ? this.Namespace : undefined
-          },
-        };
+  }
+
+  @action
+  isSecret() {
+    return this.args.type === 'secret';
+  }
+
+  @action
+  change(e) {
+    e.data.toJSON = function() {
+      return {
+        AccessorID: this.AccessorID,
+        // TODO: In the past we've always ignored the SecretID returned
+        // from the server and used what the user typed in instead, now
+        // as we don't know the SecretID when we use SSO we use the SecretID
+        // in the response
+        SecretID: this.SecretID,
+        Namespace: this.Namespace,
+        Partition: this.Partition,
+        ...{
+          AuthMethod: typeof this.AuthMethod !== 'undefined' ? this.AuthMethod : undefined,
+          // TODO: We should be able to only set namespaces if they are enabled
+          // but we might be testing for nspaces everywhere
+          // Namespace: typeof this.Namespace !== 'undefined' ? this.Namespace : undefined
+        },
       };
-      // TODO: We should probably put the component into idle state
-      this.onchange(e);
-    },
-  },
-});
+    };
+    // TODO: We should probably put the component into idle state
+    if(typeof this.args.onchange === 'function') {
```

Comment:
Do we ever use Ember Utils (i.e. https://api.emberjs.com/ember/3.28/functions/@ember%2Futils/typeOf) for things like this?

Reply:
Nah, we never use `typeOf`, partly due to this:

```javascript
console.log(typeOf(class {})); // function
console.log(typeOf(new (class {}))); // object
console.log(typeOf(EmberObject.extend())); // class
console.log(typeOf(EmberObject.create())); // instance
```

I think quite a few Ember folks don't use this anymore, I'm surprised they haven't deprecated it yet.

FYI we always use the forms:

```javascript
// checking undefined-ness as opposed to just null
typeof thing === 'undefined'
// checking whether something is a function (so callable but also not undefined)
typeof thing === 'function'
// checking for an array, I would guess this method probably didn't exist when ember's typeOf was originally written
Array.isArray(thing)
```

Generally it should be 'reasonably clear' for other primitive values what types they are, and the types should never change, and off the top of my head I don't think we ever do any other types of runtime type checking anywhere. In the past I've been very careful when writing things to make sure I'm pretty strict with this sort of stuff.

Bit of a tangent, but at one point I was running `tsc` on the project to ensure this kind of stuff (but importantly not using Typescript itself), actually theres an RFC about this somewhere and I wouldn't mind picking that up again and getting it committed to the project in some way. Another one to chat through at some point. Will link you the RFC offline.